### PR TITLE
Deprecate implicit .U() and .S() syntax for literal bit extracts

### LINKED ISSUE #2413: Isn't .U syntax supposed to only take .W arguments?
**Type of issue**: bug report

**Impact**: API modification

**Development Phase**: request 

**Other information**
 

**If the current behavior is a bug, please provide the steps to reproduce the problem:**

Put this in a `Module`:

```  
val bar = 5
val foo = bar.U(7)
```

Here is a simple Scastie (some commented out code that also all works unexpectedly)
https://scastie.scala-lang.org/nqEQGHEUTziPKNj7NUirBA

**What is the current behavior?**

This code compiles without error

**What is the expected behavior?**

I thought it should be an error to have no `.W` suffix on the arguments inside the `.U`, since those are not supposed to directly take integer arguments

**Please tell us about your environment:**

    - version: 3.5.1
    
**What is the use case for changing the behavior?**

Hard-to-catch unintended user errors

### PR DESCRIPTION
**Pull request by jared-barocsi**
Fixes #2413: It is currently possible to pass an integer to the literal .U/.S syntax as a way of immediately bit extracting on a literal value, but this poses confusion and ambiguity as users usually intend to pass a Width.

### Contributor Checklist

- [ ] Did you add Scaladoc to every public function/method?
- [ ] Did you add at least one test demonstrating the PR?
- [x] Did you delete any extraneous printlns/debugging code?
- [x] Did you specify the type of improvement?
- [ ] Did you add appropriate documentation in `docs/src`?
- [x] Did you state the API impact?
- [x] Did you specify the code generation impact?
- [x] Did you request a desired merge strategy?
- [x] Did you add text to be included in the Release Notes for this change?

#### Type of Improvement

- bug fix
- new feature/API

#### API Impact

Deprecates `.U(<int>)` and `.S(<int>)` as a way of bit extracting on a literal value, instead advising users to explicitly invoke `.U.apply(int)`

#### Backend Code Generation Impact

No impact

#### Desired Merge Strategy

Squash and merge

#### Release Notes
Deprecate implicit literal bit extracts, e.g. `1.U(8)`, to prevent users from accidentally passing an integer instead of a Width.
Instead, if a bit extract on a literal is intentional then users are advised to first assign the literal to a `val` and then extract from that to be clear about which functionality they are using.

### Reviewer Checklist (only modified by reviewer)
- [ ] Did you add the appropriate labels?
- [x] Did you mark the proper milestone (Bug fix: `3.4.x`, [small] API extension: `3.5.x`, API modification or big change: `3.6.0`)?
- [x] Did you review?
- [ ] Did you check whether all relevant Contributor checkboxes have been checked?
- [x] Did you mark as `Please Merge`?

### CODE REVIEW

**Comment by aswaterman** [comment]
Please consider that UInt.extract conflicts with a rocket-chip PML version of same: https://github.com/chipsalliance/rocket-chip/blob/1f05ea88754b421684face06e52260848d27520f/src/main/scala/util/package.scala#L151

I know we rocket-chip folks don’t get to govern Chisel API decisions unilaterally, but if you can find a way to sidestep this collision, please do. One option is just having users explicitly invoke the apply method.

**Comment by ekiwi** [comment]
> Please consider that UInt.extract conflicts with a rocket-chip PML version of same: https://github.com/chipsalliance/rocket-chip/blob/1f05ea88754b421684face06e52260848d27520f/src/main/scala/util/package.scala#L151
> 
> I know we rocket-chip folks don’t get to govern Chisel API decisions unilaterally, but if you can find a way to sidestep this collision, please do. One option is just having users explicitly invoke the apply method.

I wonder if we could just change the behavior of `apply(hi, lo)` in chisel to match the one of the `extract` method in rocket chip and then remove the `extract` from rocket chip? Is there a reason the current behavior of `apply(hi, lo)` is better?

**Comment by jackkoenig** [comment]
> I wonder if we could just change the behavior of `apply(hi, lo)` in chisel to match the one of the `extract` method in rocket chip and then remove the `extract` from rocket chip? Is there a reason the current behavior of `apply(hi, lo)` is better?

The Chisel version errors where the Rocket-Chip returns `0.U`, when `hi == lo-1`. I think it's reasonable to have the rocket-chip behavior; however it would be much better if we could return a true zero-width wire rather than a 1-bit zero. This is perhaps in the realm of "reserving the right to have that behavior" but we really need true zero-width wire support.

**Comment by jackkoenig** [comment]
A reasonable plan here is to just tell users to use `.apply` and not take the name `extract`. Some day when we have zero-width wire support we can adopt the rocket-chip behavior and add `extract` as an alias for `apply`.

Given the addition of `extract` in this PR is just to slightly increase user understandability, I think avoiding the PML collision is a fine reason to just not do it [yet] and just tell users to call `.apply`.

File: core/src/main/scala/chisel3/Bits.scala [hunk 1]
```diff
@@ -177,6 +175,12 @@ sealed abstract class Bits(private[chisel3] val width: Width) extends Element wi
     }
   }
 
+  final def apply(x: Int, y: Int): UInt = macro SourceInfoTransform.xyArg
```
Comment (jackkoenig) [inline core/src/main/scala/chisel3/Bits.scala:178]: We may end up not adding `extract`, but if we do, make sure not to get rid of ScalaDoc. All public methods need ScalaDoc.

File: core/src/main/scala/chisel3/package.scala [hunk 2]
```diff
@@ -37,13 +39,48 @@ package object chisel3 {
...
-    /** Int to SInt conversion, recommended style for constants.
-      */
-    def S: SInt = SInt.Lit(bigint, Width())
+    /** BigInt to UInt conversion followed by bit extract (bad style, deprecated) */
+    @chiselRuntimeDeprecated
+    @deprecated(
+      "Passing an Int to .U is usually a mistake: It does *not* set the width but does a bit extract. " +
+        "Did you mean .U(<arg>.W)? " +
+        "If you do want bit extraction, use .U.extract(<arg>)",
+      "3.5.3"
```
Comment (jackkoenig) [inline core/src/main/scala/chisel3/package.scala:51]:
```suggestion
      "Chisel 3.5.4"
```
You always need to include the name of the project in the version ([see ScalaDoc for `@deprecated` which says this](https://www.scala-lang.org/api/2.13.8/scala/deprecated.html)).

Also Chisel 3.5.3 has already been released so this is Chisel 3.5.4.

File: core/src/main/scala/chisel3/package.scala [hunk 3]
```diff
@@ -37,13 +39,48 @@ package object chisel3 {
...
+      // Note that both source locator and compile options are ignored for apply on literal
+      this.U.extract(x)(DeprecatedSourceInfo, ExplicitCompileOptions.Strict)
+    }
+
+    /** BigInt to UInt conversion followed by bit extract */
+    def U(x: BigInt, y: BigInt): UInt = { // scalastyle:ignore method.name
+      // Note that both source locator and compile options are ignored for apply on literal
+      this.U.do_apply(x, y)(DeprecatedSourceInfo, ExplicitCompileOptions.Strict)
+    }
+
```
Comment (jackkoenig) [inline core/src/main/scala/chisel3/package.scala:63]: Should this version also be deprecated?

**Comment by mwachs5** [comment]
> Given the addition of extract in this PR is just to slightly increase user understandability, I think avoiding the PML collision is a fine reason to just not do it [yet] and just tell users to call .apply.

Is this particularly for the 2-argument `hi`, `lo` case or are you suggesting getting rid of the `.extract()` option and suggestion for even the single argument case?

File: src/test/scala/chiselTests/UIntOps.scala [hunk 4]
```diff
@@ -172,15 +172,15 @@ class MatchedRotateLeftAndRight(w: Int = 13) extends BasicTester {
...
+  assert("b101010".U.extract(2) === false.B)
+  assert("b101010".U.extract(3) === true.B)
+  assert("b101010".U.extract(100) === false.B)
   assert("b101010".U(3, 0) === "b1010".U)
   assert("b101010".U(9, 0) === "b0000101010".U)
 
-  assert("b101010".U(6.W)(2) === false.B)
-  assert("b101010".U(6.W)(3) === true.B)
-  assert("b101010".U(6.W)(100) === false.B)
+  assert("b101010".U(6.W).extract(2) === false.B)
```
Comment (mwachs5) [inline src/test/scala/chiselTests/UIntOps.scala:181]: Why did these 3 tests need to be changed? I thought that if you put the (6.W) it *is* OK to put the extraction index immediately?

**Comment by mwachs5** [comment]
@jared-barocsi can you change the PR title and release notes to reflect this final version, as there is no longer an `.extract`

File: core/src/main/scala/chisel3/package.scala (hunk 3 above)
Comment (mwachs5) [reply core/src/main/scala/chisel3/package.scala:63]: It sounds like we agreed this version does not need to be deprecated? As we seem to think it is not ambiguous?

File: src/test/scala/chiselTests/SIntOps.scala [hunk 5]
```diff
@@ -85,11 +85,11 @@ class SIntOpsTester(c: SIntOps) extends Tester(c) {
...
 class SIntLitExtractTester extends BasicTester {
-  assert(-5.S(1) === true.B)
-  assert(-5.S(2) === false.B)
-  assert(-5.S(100) === true.B)
-  assert(-5.S(3, 0) === "b1011".U)
-  assert(-5.S(9, 0) === "b1111111011".U)
+  assert(-5.S.apply(1) === true.B)
+  assert(-5.S.apply(2) === false.B)
+  assert(-5.S.apply(100) === true.B)
+  assert(-5.S.apply(3, 0) === "b1011".U)
```
Comment (mwachs5) [inline src/test/scala/chiselTests/SIntOps.scala:91]: this is currently not consistent with the implementation -- you did not deprecate this version so you should not need the `apply` for the 2-argument version. Which way do you want to go?

File: src/test/scala/chiselTests/UIntOps.scala [hunk 6]
```diff
@@ -172,11 +172,11 @@ class MatchedRotateLeftAndRight(w: Int = 13) extends BasicTester {
...
 class UIntLitExtractTester extends BasicTester {
-  assert("b101010".U(2) === false.B)
-  assert("b101010".U(3) === true.B)
-  assert("b101010".U(100) === false.B)
-  assert("b101010".U(3, 0) === "b1010".U)
-  assert("b101010".U(9, 0) === "b0000101010".U)
+  assert("b101010".U.apply(2) === false.B)
+  assert("b101010".U.apply(3) === true.B)
+  assert("b101010".U.apply(100) === false.B)
+  assert("b101010".U.apply(3, 0) === "b1010".U)
```
Comment (mwachs5) [inline src/test/scala/chiselTests/UIntOps.scala:178]: same comment as above -- inconsistent with the implementation to add the `.apply` here

File: core/src/main/scala/chisel3/package.scala (hunk 3 above)
Comment (jackkoenig) [reply core/src/main/scala/chisel3/package.scala:63]: 👍

**Comment by jackkoenig** [comment]
> > Given the addition of extract in this PR is just to slightly increase user understandability, I think avoiding the PML collision is a fine reason to just not do it [yet] and just tell users to call .apply.
> 
> Is this particularly for the 2-argument `hi`, `lo` case or are you suggesting getting rid of the `.extract()` option and suggestion for even the single argument case?

For consistency reasons, I think we should only add `extract` if we're adding it for both versions.

File: src/test/scala/chiselTests/SIntOps.scala (hunk 5 above)
Comment (mwachs5) [reply src/test/scala/chiselTests/SIntOps.scala:91]: or maybe we should have tests for both versions

**Review by jackkoenig — CHANGES_REQUESTED** [review]
So unfortunately this doesn't currently work.

Given the following:
```scala
class Example extends Module {
  val out = IO(Output(UInt()))
  out := 0.U(8)
}
```
No warning is given, it isn't calling the new deprecated function, it just calls the normal `.U` and then `.apply` on the returned `UInt`.

If you manually pass a `BigInt` you get:
```scala
class Example extends Module {
  val out = IO(Output(UInt()))
  out := 0.U(BigInt(8))
}
```
This gives the following error:
```
[error] .../Jack.scala:6:12: ambiguous reference to overloaded definition,
[error] both method U in class fromBigIntToLiteral of type (x: BigInt)chisel3.Bool
[error] and  method U in class fromBigIntToLiteral of type => chisel3.UInt
[error] match argument types (scala.math.BigInt) and expected result type chisel3.Data
[error]   out := 0.U(BigInt(8))
[error]            ^
```

I'm not sure what exactly we need to do to make this work, but I'm guessing it involves turning `def U` into a macro.

File: core/src/main/scala/chisel3/Bits.scala [hunk 7]
```diff
@@ -94,9 +100,9 @@ sealed abstract class Bits(private[chisel3] val width: Width) extends Element wi
     * @param x an index
     * @return the specified bit
     */
-  final def apply(x: BigInt): Bool = macro SourceInfoTransform.xArg
+  final def apply(x: BigInt): Bool = macro IntLiteralApplyTransform.safeApply
 
-  /** @group SourceInfoTransformMacro */
+  /** @group IntLiteralApplyTransformMacro */
```
Comment (jackkoenig) [inline core/src/main/scala/chisel3/Bits.scala:105]:
```suggestion
  /** @group SourceInfoTransformMacro */
```
This group is not related to the actual `macro` class, it's purely used for ScalaDoc grouping, leave that bit alone.

File: core/src/main/scala/chisel3/Bits.scala [hunk 8]
```diff
@@ -121,12 +127,12 @@ sealed abstract class Bits(private[chisel3] val width: Width) extends Element wi
...
     * @param x an index
     * @return the specified bit
-    * @note convenience method allowing direct use of [[scala.Int]] without implicits
     */
-  final def apply(x: Int): Bool = macro SourceInfoTransform.xArg
+  final def apply(x: Int): Bool = macro IntLiteralApplyTransform.safeApply
 
-  /** @group SourceInfoTransformMacro */
-  final def do_apply(x: Int)(implicit sourceInfo: SourceInfo, compileOptions: CompileOptions): Bool = apply(BigInt(x))
+  /** @group IntLiteralApplyTransformMacro */
```
Comment (jackkoenig) [inline core/src/main/scala/chisel3/Bits.scala:133]:
```suggestion
  /** @group SourceInfoTransformMacro */
```
This group is not related to the actual `macro` class, it's purely used for ScalaDoc grouping, leave that bit alone.

File: macros/src/main/scala/chisel3/internal/sourceinfo/SourceInfoTransform.scala [hunk 9]
```diff
@@ -294,3 +293,33 @@ class SourceInfoWhiteboxTransform(val c: whitebox.Context) extends AutoSourceTra
     q"$thisObj.$doFuncTerm($that)($implicitSourceInfo, $implicitCompileOptions)"
   }
 }
+
+// Workaround for https://github.com/sbt/sbt/issues/3966
+object IntLiteralApplyTransform
+
+/**
+  */
```
Comment (jackkoenig) [inline macros/src/main/scala/chisel3/internal/sourceinfo/SourceInfoTransform.scala:301]:
```suggestion
```
No purpose in adding an empty ScalaDoc

**Review by jackkoenig — CHANGES_REQUESTED** [review]
This is awesome, works great. I have some minor suggestions but I have checked this out locally and it works great! I couldn't find a way to test for compiler warnings in ScalaTest so I think we'll just go with it as is.

**Review by jackkoenig — APPROVED** [review]
LGTM!

**Pull request merged**